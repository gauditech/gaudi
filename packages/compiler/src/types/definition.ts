import { FieldType, TypeCardinality } from "@compiler/compiler/ast/type";
import { HookCode } from "@compiler/types/common";
import { Literal } from "@compiler/types/specification";

export type Definition = {
  validators: ValidatorDef[];
  models: ModelDef[];
  apis: ApiDef[];
  populators: PopulatorDef[];
  runtimes: ExecutionRuntimeDef[];
  authenticator: AuthenticatorDef | undefined;
  generators: GeneratorDef[];
};

export type ValidatorDef = {
  name: string;
  args: { name: string; type: FieldType }[];
  assert: TypedExprDef;
  error: { code: string };
};

export type ValidatorHookDef = {
  hook: HookCode;
  args: { name: string; expr: TypedExprDef }[];
};

export type ValidateExprDef =
  | { kind: "and" | "or"; exprs: ValidateExprDef[] }
  | ValidateExprCallDef;
export type ValidateExprCallDef = { kind: "call"; validator: string; args: TypedExprDef[] };

export type ModelDef = {
  kind: "model";
  refKey: string;
  name: string;
  dbname: string;
  fields: FieldDef[];
  references: ReferenceDef[];
  relations: RelationDef[];
  queries: QueryDef[];
  aggregates: AggregateDef[];
  computeds: ComputedDef[];
  hooks: ModelHookDef[];
};

export type FieldDef = {
  kind: "field";
  refKey: string;
  modelRefKey: string;
  name: string;
  dbname: string;
  type: FieldType;
  primary: boolean;
  unique: boolean;
  nullable: boolean;
  validate?: ValidateExprDef;
};

export type ReferenceDef = {
  kind: "reference";
  refKey: string;
  modelRefKey: string;
  name: string;
  fieldRefKey: string;
  toModelRefKey: string;
  toModelFieldRefKey: string;
  nullable: boolean;
  unique: boolean;
  onDelete?: ReferenceOnDeleteAction;
};
export type ReferenceOnDeleteAction = "setNull" | "cascade";

export type RelationDef = {
  kind: "relation";
  refKey: string;
  modelRefKey: string;
  name: string;
  fromModel: string;
  fromModelRefKey: string;
  through: string;
  throughRefKey: string;
  unique: boolean;
};

export type QueryDef = {
  kind: "query";
  refKey: string;
  modelRefKey: string;
  name: string;
  // retType: string | "integer";
  retType: string;
  retCardinality: TypeCardinality;
  fromPath: string[];
  // unique: boolean;
  filter: TypedExprDef;
  select: SelectDef;
  orderBy: QueryOrderByAtomDef[] | undefined;
  limit: number | undefined;
  offset: number | undefined;
};

export type QueryOrderByAtomDef = { exp: TypedExprDef; direction: "asc" | "desc" };

export type AggregateDef = {
  kind: "aggregate";
  refKey: string;
  name: string;
  aggrFnName: "count" | "sum";
  targetPath: string[];
  query: Omit<QueryDef, "refKey" | "select" | "name">;
};

export type ComputedDef = {
  kind: "computed";
  refKey: string;
  modelRefKey: string;
  name: string;
  exp: TypedExprDef;
  type: VariablePrimitiveType;
};

export type ModelHookDef = {
  kind: "model-hook";
  refKey: string;
  name: string;
  args: { name: string; query: QueryDef }[];
  hook: HookCode;
};

export type VariablePrimitiveType = {
  kind: FieldType | "null";
  nullable: boolean;
};

type VariableCollectionType = {
  kind: "collection";
  type: VariablePrimitiveType;
};

type TypedVariableType = VariablePrimitiveType | VariableCollectionType;

export type LiteralValueDef = { kind: "literal"; literal: Literal };

type TypedAlias = { kind: "alias"; namePath: string[]; type?: TypedVariableType };
type TypedVariable = { kind: "variable"; type?: TypedVariableType; name: string };

export type BinaryOperator =
  | "or"
  | "and"
  | "is not"
  | "is"
  | "not in"
  | "in"
  | "<"
  | "<="
  | ">"
  | ">="
  | "+"
  | "-"
  | "/"
  | "*";

export type FunctionName =
  | BinaryOperator
  | "length"
  | "concat"
  | "lower"
  | "upper"
  | "now"
  | "cryptoHash"
  | "cryptoCompare"
  | "cryptoToken"
  | "stringify";

export type AggregateFunctionName = "count" | "sum";

export type InCollectionFunctionName = "in" | "not in";

export type TypedFunction = {
  kind: "function";
  name: FunctionName;
  args: TypedExprDef[];
  type?: TypedVariableType;
};

export type TypedExprDef =
  | LiteralValueDef
  | TypedArray
  | TypedAlias
  | TypedVariable
  | TypedFunction
  | TypedAggregateFunction
  | TypedExistsSubquery
  | undefined;

type TypedArray = {
  kind: "array";
  elements: TypedExprDef[];
  type: VariableCollectionType;
};

type TypedAggregateFunction = {
  kind: "aggregate-function";
  fnName: FunctionName | AggregateFunctionName;
  type: VariablePrimitiveType;
  sourcePath: string[];
  targetPath: string[];
};

type TypedExistsSubquery = {
  kind: "in-subquery";
  fnName: InCollectionFunctionName;
  lookupExpression: TypedExprDef;
  sourcePath: string[];
  targetPath: string[];
};

export type ApiDef = {
  name?: string;
  path: string;
  entrypoints: EntrypointDef[];
};

/**
 * ENTRYPOINTS
 */
export type EntrypointDef = {
  name: string;
  target: TargetDef;
  endpoints: EndpointDef[];
  entrypoints: EntrypointDef[];
};

export type TargetDef = {
  kind: "model" | "reference" | "relation";
  name: string;
  namePath: string[];
  retType: string;
  alias: string;
  identifyWith:
    | {
        path: string[];
        // FIXME we should support any field type that can be represented as a string
        // as long as there's a unique index
        type: "string" | "integer";
        paramName: string;
      }
    | undefined;
};

export type TargetWithSelectDef = TargetDef & { select: SelectDef };

export type EndpointHttpMethod = "GET" | "POST" | /*"PUT" |*/ "PATCH" | "DELETE";

export type EndpointDef =
  | CreateEndpointDef
  | ListEndpointDef
  | GetEndpointDef
  | UpdateEndpointDef
  | DeleteEndpointDef
  | CustomOneEndpointDef
  | CustomManyEndpointDef;

export type ListEndpointDef = {
  kind: "list";
  parentContext: TargetWithSelectDef[];
  target: Omit<TargetWithSelectDef, "identifyWith">;
  authSelect: SelectDef;
  authorize: TypedExprDef;
  pageable: boolean;
  response: SelectDef;
  fieldset?: undefined;
  orderBy: QueryOrderByAtomDef[] | undefined;
  filter: TypedExprDef | undefined;
  // actions: ActionDef[];
};

export type GetEndpointDef = {
  kind: "get";
  parentContext: TargetWithSelectDef[];
  target: TargetWithSelectDef;
  authSelect: SelectDef;
  authorize: TypedExprDef;
  response: SelectDef;
  fieldset?: undefined;
  // actions: ActionDef[];
};

export type CreateEndpointDef = {
  kind: "create";
  parentContext: TargetWithSelectDef[];
  target: Omit<TargetWithSelectDef, "identifyWith">;
  authSelect: SelectDef;
  authorize: TypedExprDef;
  response: SelectDef;
  fieldset?: FieldsetDef;
  actions: ActionDef[];
};

export type UpdateEndpointDef = {
  kind: "update";
  parentContext: TargetWithSelectDef[];
  target: TargetWithSelectDef;
  authSelect: SelectDef;
  authorize: TypedExprDef;
  response: SelectDef;
  fieldset?: FieldsetDef;
  actions: ActionDef[];
};

export type DeleteEndpointDef = {
  kind: "delete";
  parentContext: TargetWithSelectDef[];
  target: TargetWithSelectDef;
  actions: ActionDef[];
  authSelect: SelectDef;
  authorize: TypedExprDef;
  response: undefined;
  fieldset?: undefined;
};

export type CustomOneEndpointDef = {
  kind: "custom-one";
  parentContext: TargetWithSelectDef[];
  target: TargetWithSelectDef;
  method: EndpointHttpMethod;
  path: string;
  actions: ActionDef[];
  authSelect: SelectDef;
  authorize: TypedExprDef;
  fieldset?: FieldsetDef;
  response: undefined;
  responds: boolean;
};

export type CustomManyEndpointDef = {
  kind: "custom-many";
  parentContext: TargetWithSelectDef[];
  target: Omit<TargetWithSelectDef, "identifyWith">;
  method: EndpointHttpMethod;
  path: string;
  actions: ActionDef[];
  authSelect: SelectDef;
  authorize: TypedExprDef;
  fieldset?: FieldsetDef;
  response: undefined;
  responds: boolean;
};

export type SelectableExpression = {
  kind: "expression";
  alias: string;
  expr: TypedExprDef;
  type: VariablePrimitiveType;
};

/**
 * NOTE: hooks currently cannot be part of expressions.
 */

export type SelectHook = {
  kind: "model-hook";
  refKey: string;
  name: string;
  alias: string;
  namePath: string[];
};

export type NestedSelect = {
  kind: "nested-select";
  refKey: string;
  namePath: string[];
  alias: string;
  select: SelectItem[];
};

export type SelectItem = SelectableExpression | NestedSelect | SelectHook;

export type SelectDef = SelectItem[];

export type FieldsetDef = FieldsetRecordDef | FieldsetFieldDef;

export type FieldsetRecordDef = {
  kind: "record";
  record: Record<string, FieldsetDef>;
  nullable: boolean;
};

export type IValidatorDef = {
  name: string;
  inputType: FieldType;
  args: Literal[];
};

export type FieldsetFieldDef = {
  kind: "field";
  type: FieldType;
  nullable: boolean;
  required: boolean;
  validate?: ValidateExprDef;
  referenceNotFound?: true;
};

export interface ReferenceNotFoundValidator {
  name: "reference-not-found";
}

export type ActionDef =
  | CreateOneAction
  | UpdateOneAction
  | DeleteOneAction
  | ExecuteHookAction
  | QueryAction
  | RespondAction
  | ValidateAction;

export type CreateOneAction = {
  kind: "create-one";
  alias: string;
  model: string;
  targetPath: string[];
  changeset: ChangesetDef;
  select: SelectDef;
  isPrimary: boolean;
};

export type UpdateOneAction = {
  kind: "update-one";
  alias: string;
  model: string;
  targetPath: string[];
  filter: TypedExprDef;
  changeset: ChangesetDef;
  select: SelectDef;
  isPrimary: boolean;
};

export type DeleteOneAction = {
  kind: "delete-one";
  model: string;
  targetPath: string[];
};

export type ExecuteHookAction = {
  kind: "execute-hook";
  alias: string;
  hook: ActionHookDef;
  responds: boolean;
};

export type RespondAction = {
  kind: "respond";
  body: FieldSetter;
  httpStatus?: FieldSetter;
  httpHeaders?: { name: string; value: FieldSetter }[];
};

export type QueryAction = {
  kind: "query";
  alias: string;
  model: string;
  query: QueryDef;
};

export type ValidateAction = {
  kind: "validate";
  key: string;
  validate: ValidateExprDef;
};

export type ActionHookDef = {
  hook: HookCode;
  args: ChangesetDef;
};

export type ChangesetDef = ChangesetOperationDef[];
export type ChangesetOperationDef = { name: string; setter: FieldSetter };

export type FieldSetterReferenceValue = {
  kind: "reference-value";
  target: { alias: string; access: string[] };
};

export type FieldSetterInput = {
  kind: "fieldset-input";
  type: FieldType;
  fieldsetAccess: string[];
  required: boolean;
  // FIXME implement default
  // default?: LiteralValueDef | FieldSetterReferenceValue;
};

export type FieldSetterReferenceInput = {
  kind: "fieldset-reference-input";
  fieldsetAccess: string[];
  through: string[];
  // required: boolean;
};

export type PopulatorDef = {
  name: string;
  populates: PopulateDef[];
};

export type PopulateDef = {
  target: TargetDef;
  actions: ActionDef[];
  populates: PopulateDef[];
  repeater: RepeaterDef;
};

export type RepeaterDef = { alias?: string; start: number; end: number };

export type FieldSetterChangesetReference = {
  kind: "changeset-reference";
  referenceName: string;
};

export type FieldSetterHttpHandler = {
  kind: "request-auth-token";
  access: string[];
};

export type FieldSetterFunction = {
  kind: "function";
  name: FunctionName; // TODO rename to `fnName` to make it more clear, see line 124 as well
  args: FieldSetter[];
};

export type FieldSetterContextReference = {
  kind: "context-reference";
  referenceName: string;
};

export type FieldSetterHook = {
  kind: "fieldset-hook";
  hook: HookCode;
  args: ChangesetDef;
};

export type FieldSetterQuery = {
  kind: "query";
  query: QueryDef;
};

export type FieldSetterArray = {
  kind: "array";
  elements: FieldSetter[];
};

export type FieldSetter =
  // TODO add composite expression setter
  | LiteralValueDef
  | FieldSetterReferenceValue
  | FieldSetterInput
  | FieldSetterReferenceInput
  | FieldSetterChangesetReference
  | FieldSetterHook
  | FieldSetterHttpHandler
  | FieldSetterFunction
  | FieldSetterContextReference
  | FieldSetterQuery
  | FieldSetterArray;

export type ExecutionRuntimeDef = {
  name: string;
  type: RuntimeEngineType;
  sourcePath: string;
};

export type RuntimeEngineType = "node";

// ---------- authenticator

export type AuthenticatorDef = {
  name: string;
  authUserModel: AuthenticatorNamedModelDef;
  accessTokenModel: AuthenticatorNamedModelDef;
  method: AuthenticatorMethodDef;
};

export type AuthenticatorNamedModelDef = {
  name: string;
  refKey: string;
};

export type AuthenticatorMethodDef = AuthenticatorBasicMethodDef;

export type AuthenticatorBasicMethodDef = {
  kind: "basic";
};

// ----- Generators

export type GeneratorDef = GeneratorClientDef;

export type GeneratorClientTarget = "js" | "ts";

export type GeneratorClientDef = {
  kind: "generator-client";
  target: GeneratorClientTarget;
  output?: string;
};
