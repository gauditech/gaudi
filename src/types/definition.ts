export type Definition = {
  models: ModelDef[];
};

export type ModelDef = {
  refKey: string;
  name: string;
  dbname: string;
  fields: FieldDef[];
  references: ReferenceDef[];
  relations: RelationDef[];
  queries: QueryDef[];
};

export type FieldDef = {
  refKey: string;
  modelRefKey: string;
  name: string;
  dbname: string;
  type: "integer" | "text" | "boolean";
  dbtype: "serial" | "integer" | "text" | "boolean";
  primary: boolean;
  unique: boolean;
  nullable: boolean;
};

export type ReferenceDef = {
  refKey: string;
  modelRefKey: string;
  name: string;
  fieldRefKey: string;
  toModelRefKey: string;
  toModelFieldRefKey: string;
  nullable: boolean;
  unique: boolean;
};

export type RelationDef = {
  refKey: string;
  modelRefKey: string;
  name: string;
  fromModel: string;
  fromModelRefKey: string;
  through: string;
  throughRefKey: string;
  nullable: boolean;
  unique: boolean;
};

type QueryDefPathSelect = {
  name: string;
  retType: string;
  nullable: boolean;
  refKey: string;
};

export type QueryDefPath = {
  refKey: string;
  name: string;
  retType: string;
  retCardinality: "one" | "many"; // should be retCard,...
  nullable: boolean;
  joinType: "inner" | "left";
  alias: string;
  bpAlias: string | null;
  joinPaths: QueryDefPath[];
  select: QueryDefPathSelect[];
};

type QueryDefFilter = {
  type: string;
  lhs: string | number | boolean;
  rhs: string | number | boolean;
};

export type QueryDef = {
  refKey: string;
  name: string;
  retType: string;
  retCardinality: "one" | "many";
  nullable: boolean;
  // unique: boolean;
  joinPaths: QueryDefPath[];
  filters: QueryDefFilter[];
};
