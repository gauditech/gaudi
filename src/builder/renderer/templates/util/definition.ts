import { Definition, PathFragment } from "@src/types/definition";

/*
 * Template helper functions for Gaudi definition
 */

export function modelDbName(modelRefKey: string, definition: Definition) {
  return (definition.models || []).find((model) => model.refKey === modelRefKey)?.dbname;
}

export function fieldDbName(modelRefKey: string, fieldRefKey: string, definition: Definition) {
  return (definition.models || [])
    .find((model) => model.refKey === modelRefKey)
    ?.fields.find((field) => field.refKey === fieldRefKey)?.dbname;
}

export function typeToDbType(type: string) {
  if (type === "serial") {
    return "Int";
  } else if (type === "integer") {
    return "Int";
  } else if (type === "text") {
    return "String";
  } else if (type === "boolean") {
    return "Boolean";
  }

  throw "UNHANDLED_GAUDI_TYPE";
}

export function buildEndpointPath(parts: PathFragment[]): string {
  return parts.reduce((accum, part) => {
    const pathPart =
      part.type === "numeric" || part.type === "text" ? `:${part.varname}` : part.value;
    return `${accum}/${pathPart}`;
  }, "");
}
