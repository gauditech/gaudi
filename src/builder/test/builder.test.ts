import fs from "fs";
import path from "path";

import definition from "@examples/git/definition.json";

import {
  renderDbSchema,
  renderServerEndpoints,
  renderIndex,
  renderPackage,
  renderServer,
} from "@src/builder/builder";

const SNAPSHOT_FOLDER = __dirname;

describe("builder", () => {
  describe("build package", () => {
    it("renders package template correctly", async () => {
      const data = {
        package: {
          name: "test",
          description: "Test description",
          version: "0.0.1",
        },
      };
      const snapshot = readSnapshot("package.json");

      expect(await renderPackage(data)).toEqual(snapshot);
    });
  });

  describe("build index", () => {
    it("renders index template correctly", async () => {
      const snapshot = readSnapshot("index.js");

      expect(await renderIndex()).toEqual(snapshot);
    });
  });

  describe("build DB schema", () => {
    it("renders DB schema template correctly", async () => {
      const data = {
        definition: definition as any,
        dbProvider: "DB_PROVIDER",
        dbConnectionUrl: "DB_CONNECTION_URL",
      };
      const snapshot = readSnapshot("schema.prisma");

      expect(await renderDbSchema(data)).toEqual(snapshot);
    });
  });

  describe("build server", () => {
    it("renders server template correctly", async () => {
      const data = { serverPort: 3001 };

      expect(await renderServer(data)).toMatchSnapshot();
    });

    it("renders server get endpoints correctly", async () => {
      const data = {
        definition: {
          entrypoints: [
            {
              endpoints: [
                {
                  name: "get",
                  kind: "get",
                  identifyRefPath: ["Org.id"],
                  path: [
                    { type: "literal", value: "org" },
                    { type: "numeric", varname: "org_id" },
                  ],
                  actions: [
                    {
                      kind: "fetch one",
                      modelRef: "Org",
                      filter: {
                        kind: "binary",
                        operation: "is",
                        lhs: "Org.id",
                        rhs: { kind: "var ref", varname: "org_id" },
                      },
                      select: {
                        fieldRefs: ["Org.id", "Org.name"],
                        queries: [],
                        references: [],
                        relations: [],
                      },
                      varname: "var0",
                      onError: { statusCode: 404, body: { message: "Not found" } },
                    },
                    { kind: "respond", varname: "var0" },
                  ],
                },
              ],
            },
          ],
        } as any,
      };

      expect(await renderServerEndpoints(data as any)).toMatchSnapshot();
    });

    it("renders server list endpoints correctly", async () => {
      const data = {
        definition: {
          entrypoints: [
            {
              endpoints: [
                {
                  name: "list",
                  kind: "list",
                  path: [{ type: "literal", value: "org" }],
                  actions: [
                    {
                      kind: "fetch many",
                      modelRef: "Org",
                      filter: undefined,
                      select: {
                        fieldRefs: ["Org.id", "Org.name"],
                        queries: [],
                        references: [],
                        relations: [],
                      },
                      varname: "var0",
                    },
                    { kind: "respond", varname: "var0" },
                  ],
                },
              ],
            },
          ],
        },
      };

      expect(await renderServerEndpoints(data as any)).toMatchSnapshot();
    });
  });
});

// ---------- helpers

function readSnapshot(baseFilename: string): string {
  return fs.readFileSync(path.join(SNAPSHOT_FOLDER, `${baseFilename}.snapshot`)).toString();
}
