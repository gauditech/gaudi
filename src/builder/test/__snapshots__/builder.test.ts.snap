// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder build server renders server get endpoints correctly 1`] = `
"const { endpointHandlerGuard, EndpointError } = require("./common.js");

  // setup endpoints
  const endpointConfigs = [];
  function setupEndpoints(app) {
    endpointConfigs.forEach((ep) => {
      app[ep.method](ep.path, endpointHandlerGuard(ep.handler));
    })
  } 

  // hello world endpoint
  function helloEndpoint(req, res) { res.send("Hello world!") }
  endpointConfigs.push({ path: '/hello', method: 'get', handler: helloEndpoint })

  // definition endpoints
  // getEndpoint
    async function getEndpoint(req, resp) {
      // TODO: role auth
      // TODO: select
      // TODO: filter

      const ctx = new Map();

      // extract path vars
      ctx.set("org_id", req.params["org_id"]);
          
      try {
        // actions
        try {
  const var0 = await fetchOneAction(ctx);
  ctx.set("var0", var0)
}
catch(err) {
  throw new EndpointError(404, {"message":"Not found"}, err)
}
        resp.send(ctx.get("var0"))
      } catch(err) {
        if (err instanceof EndpointError) {
          throw err;
        }
        else {
          throw new EndpointError(500, 'Error processing request: ' + err);
        }
      }
    }
    endpointConfigs.push({ path: "/org/:org_id", method: "get", handler: getEndpoint })

  // ----- commons
  
  async function fetchOneAction(ctx) {
    // example endpoint error
    const orgId = ctx.get('org_id');
    if (orgId == 3) {
      throw new Error('Example action error')
    }

    // TODO: execute query

    return \`Example fetch-one action result: \${orgId}\`
  }

  async function fetchManyAction(ctx) {
    // TODO: execute query
    return ['Example list action result']
  }

  // ----- export

  module.exports = {
    setupEndpoints
  }"
`;

exports[`builder build server renders server list endpoints correctly 1`] = `
"const { endpointHandlerGuard, EndpointError } = require("./common.js");

  // setup endpoints
  const endpointConfigs = [];
  function setupEndpoints(app) {
    endpointConfigs.forEach((ep) => {
      app[ep.method](ep.path, endpointHandlerGuard(ep.handler));
    })
  } 

  // hello world endpoint
  function helloEndpoint(req, res) { res.send("Hello world!") }
  endpointConfigs.push({ path: '/hello', method: 'get', handler: helloEndpoint })

  // definition endpoints
  // listEndpoint
    async function listEndpoint(req, resp) {
      // TODO: role auth
      // TODO: select
      // TODO: filter
    
      const ctx = new Map();

      
          
      try {
        const var0 = await fetchManyAction(ctx);
ctx.set("var0", var0)
        resp.send(ctx.get("var0"))
      } catch(err) {
        if (err instanceof EndpointError) {
          throw err;
        }
        else {
          throw new EndpointError(500, 'Error processing request: ' + err);
        }
      }
    }
    endpointConfigs.push({ path: "/org", method: "get", handler: listEndpoint })

  // ----- commons
  
  async function fetchOneAction(ctx) {
    // example endpoint error
    const orgId = ctx.get('org_id');
    if (orgId == 3) {
      throw new Error('Example action error')
    }

    // TODO: execute query

    return \`Example fetch-one action result: \${orgId}\`
  }

  async function fetchManyAction(ctx) {
    // TODO: execute query
    return ['Example list action result']
  }

  // ----- export

  module.exports = {
    setupEndpoints
  }"
`;

exports[`builder build server renders server template correctly 1`] = `
"const express = require("express");
const { requestLogger, errorLogger, errorResponder } = require("./common.js");
const { setupEndpoints } = require("./endpoints.js");

const app = express();
const port = 3001;

setupEndpoints(app);

app.use(requestLogger)
app.use(errorLogger)
app.use(errorResponder)

app.listen(port, () => {
  console.log(\`Example app listening on port \${port}\`);
});"
`;
